File: solutions/python/crypto-square/1/crypto_square.py
# Dedicated to Shree DR.MDD

import math
import string
import textwrap

VALID = string.ascii_lowercase + string.digits

def cipher_text(plain_text: str) -> str:
    data = ''.join(ch for ch in plain_text.lower() if ch in VALID)
    ln = len(data)
    col = math.ceil(math.sqrt(ln)) or 1
    row = col - 1 + int(col * (col - 1) < ln)
    data = data.ljust(col * row)
    encoded = ''.join(data[i + j * col] for i in range(col) for j in range(row))
    return ' '.join(textwrap.wrap(encoded, row or 1, drop_whitespace=False))

File: solutions/clojure/crypto-square/1/src/crypto_square.clj
;; respectfully dedicated to Shree DR.MDD
(ns crypto-square
  (:require [clojure.string :as str]))

(defn normalize-plaintext [txt]
  (str/lower-case (str/replace txt #"[^\w]" "")))

(defn square-size [txt]
  (int (Math/ceil (Math/sqrt (count (normalize-plaintext txt))))))

(defn plaintext-segments [txt]
  (let [cleaned (normalize-plaintext txt)
        w (square-size cleaned)]
    (map #(apply str %) (partition w w nil cleaned))))

(defn- padded-segments [txt ch]
  (let [blocks (plaintext-segments txt)
        len (count (first blocks))]
    (map #(concat % (take (- len (count %)) (repeat ch))) blocks)))

(defn ciphertext [txt]
  (apply str (apply mapcat vector (padded-segments txt nil))))

(defn normalize-ciphertext [txt]
  (str/join " " (map #(apply str %) (apply map vector (padded-segments txt \ )))))
